libraries/common/list.effekt

```diff
@@ -649,35 +649,113 @@ def partition[A](l: List[A]) { pred: A => Bool }: (List[A], List[A]) = {
   (lefts.reverse, rights.reverse)
 }
 
-/// Sort a list using a given comparison function.
+/// Utilities for sorting, see 'sortBy' for more details.
+namespace sort {
+  /// Splits the given list into monotonic segments (so a list of lists).
+  ///
+  /// Internally used in the mergesort 'sortBy' to prepare the to-be-merged partitions.
+  def sequences[A](list: List[A]) { compare: (A, A) => Bool }: List[List[A]] = list match {
+    case Cons(a, Cons(b, rest)) =>
+      if (compare(a, b)) {
+        ascending(b, rest) { diffRest => Cons(a, diffRest) } {compare}
+      } else {
+        descending(b, [a], rest) {compare}
+      }
+    case _ => [list]
+  }
+
+  /// When in an ascending sequence, try to add `current` to `run` (if possible)
+  def ascending[A](current: A, rest: List[A]) { runDiff: List[A] => List[A] } { compare: (A, A) => Bool }: List[List[A]] = rest match {
+    case Cons(next, tail) and compare(current, next) =>
+      ascending(next, tail) { diffRest => runDiff(Cons(current, diffRest)) }  {compare}
+    case _ => Cons(runDiff([current]), sequences(rest) {compare})
+  }
+
+  /// When in an descending sequence, try to add `current` to `run` (if possible)
+  def descending[A](current: A, run: List[A], rest: List[A]) { compare: (A, A) => Bool }: List[List[A]] = rest match {
+    case Cons(next, tail) and not(compare(current, next)) =>
+      descending(next, Cons(current, run), tail) {compare}
+    case _ => Cons(Cons(current, run), sequences(rest) {compare})
+  }
+
+  def mergeAll[A](runs: List[List[A]]) { compare: (A, A) => Bool }: List[A] = runs match {
+    case Cons(single, Nil()) => single
+    case _ => {
+      // recursively merge in pairs until there's only a single list
+      val newRuns = mergePairs(runs) {compare}
+      mergeAll(newRuns) {compare}
+    }
+  }
+
+  def mergePairs[A](runs: List[List[A]]) { compare: (A, A) => Bool }: List[List[A]] = runs match {
+    case Cons(a, Cons(b, rest)) =>
+      Cons(merge(a, b) {compare}, mergePairs(rest) {compare})
+    case _ => runs
+  }
```

Comment:
OPT: We could make `mergeAll` iterative and use `mergePairs` as a subroutine (untested) [?]
```suggestion
  def mergeAll[A](runs: List[List[A]]) { compare: (A, A) => Bool }: List[A] = {
    def mergePairs[A](runs: List[List[A]]): List[List[A]] = runs match {
      case Cons(a, Cons(b, rest)) =>
        Cons(merge(a, b) {compare}, mergePairs(rest))
      case _ => runs
    }
    
    var currentRuns = runs
    while (currentRuns.length > 1) {
      currentRuns = mergePairs(currentRuns)
    }
    currentRuns match {
      case Nil() => Nil()
      case Cons(single, _) => single
    }
  }
```

Reply:
Actually, we could make `mergePairs` a subroutine already, that might help with inlining and SAT 🤔

Reply:
or we could just inline `mergePairs` altogether...
```suggestion
  def mergeAll[A](runs: List[List[A]]) { compare: (A, A) => Bool }: List[A] = {
    var currentRuns = runs
    while (currentRuns is Cons(a, Cons(b, rest))) {
      currentRuns = Cons(merge(a, b) {compare}, rest)
    }
    currentRuns match {
      case Nil() => Nil()
      case Cons(single, _) => single
    }
  }
```

Reply:
No, that doesn't work.